barretenberg/cpp/src/barretenberg/dsl/acir_format/acir_to_constraint_buf.cpp

```diff
@@ -82,13 +123,60 @@ T deserialize_any_format(std::vector<uint8_t>&& buf,
 
-/**
- * @brief Deserializes a `Program` from bytes, trying `msgpack` or `bincode` formats.
- * @note Ignores the Brillig parts of the bytecode when using `msgpack`.
- */
-Acir::Program deserialize_program(std::vector<uint8_t>&& buf)
+AcirFormat circuit_serde_to_acir_format(Acir::Circuit const& circuit)
+{
+    AcirFormat af;
+    // `varnum` is the true number of variables, thus we add one to the index which starts at zero
+    af.varnum = circuit.current_witness_index + 1;
+    af.num_acir_opcodes = static_cast<uint32_t>(circuit.opcodes.size());
+    af.public_inputs = join({ transform::map(circuit.public_parameters.value, [](auto e) { return e.value; }),
+                              transform::map(circuit.return_values.value, [](auto e) { return e.value; }) });
+    // Map to a pair of: BlockConstraint, and list of opcodes associated with that BlockConstraint
+    // NOTE: We want to deterministically visit this map, so unordered_map should not be used.
+    std::map<uint32_t, std::pair<BlockConstraint, std::vector<size_t>>> block_id_to_block_constraint;
+    for (size_t i = 0; i < circuit.opcodes.size(); ++i) {
+        const auto& gate = circuit.opcodes[i];
+        std::visit(
+            [&](auto&& arg) {
+                using T = std::decay_t<decltype(arg)>;
+                if constexpr (std::is_same_v<T, Acir::Opcode::AssertZero>) {
+                    handle_arithmetic(arg, af, i);
+                } else if constexpr (std::is_same_v<T, Acir::Opcode::BlackBoxFuncCall>) {
+                    handle_blackbox_func_call(arg, af, i);
+                } else if constexpr (std::is_same_v<T, Acir::Opcode::MemoryInit>) {
+                    auto block = handle_memory_init(arg);
+                    uint32_t block_id = arg.block_id.value;
+                    block_id_to_block_constraint[block_id] = { block, /*opcode_indices=*/{ i } };
+                } else if constexpr (std::is_same_v<T, Acir::Opcode::MemoryOp>) {
+                    auto block = block_id_to_block_constraint.find(arg.block_id.value);
+                    if (block == block_id_to_block_constraint.end()) {
+                        throw_or_abort("unitialized MemoryOp");
+                    }
+                    handle_memory_op(arg, af, block->second.first);
+                    block->second.second.push_back(i);
+                } else if constexpr (std::is_same_v<T, Acir::Opcode::BrilligCall>) {
+                    info("ENCOUNTERED BRILLIG CALL");
+                } else {
+                    bb::assert_failure("circuit_serde_to_acir_format: Unrecognized Acir Opcode.");
+                }
+            },
+            gate.value);
+    }
+    for (const auto& [block_id, block] : block_id_to_block_constraint) {
+        // Note: the trace will always be empty for ReturnData since it cannot be explicitly read from in noir
+        if (!block.first.trace.empty() || block.first.type == BlockType::ReturnData ||
+            block.first.type == BlockType::CallData) {
```

Comment:
hmm is it clear why CallData is included here? Probably not an issue because I don't think we'll ever get CallData with an empty trace but its confusing to have it called out specifically here. Maybe someone got confused about the ReturnData/CallData difference. (Or maybe I'm the one that's confused :))